examples/Makefile

```diff
@@ -43,155 +35,117 @@ GATK_GERMLINE_FILES	=	gatk/GATK_Germline/data/ref/ref.dict \
-	pachctl create repo images
-	pachctl create pipeline -f opencv/edges.json
-	pachctl create pipeline -f opencv/montage.json
-	pachctl put file images@master -i opencv/images.txt
-	pachctl put file images@master -i opencv/images2.txt
-
-opencv-testing: 
-	pachctl create repo images
-	$(JQ) "$(STATS)" opencv/edges.json | pachctl create pipeline -f -
-	$(JQ) "$(STATS)" opencv/montage.json | pachctl create pipeline -f -
-	pachctl start transaction
-	pachctl start commit images@master
-	pachctl finish transaction
-	pachctl put file images@master -i opencv/images.txt
-	pachctl put file images@master -i opencv/images2.txt
-	pachctl put file images@master -i opencv/images3.txt
-	pachctl put file images@master -i opencv/images4.txt
-	pachctl delete file images@master:/w7RVTsv.jpg
-	pachctl finish commit images@master
-
-opencv-testing-delete:
-	-pachctl delete pipeline montage
-	-pachctl delete pipeline edges
-
-opencv-delete: opencv-testing-delete
-	-pachctl delete repo images
-
-ml/object-detection/frozen_inference_graph.pb:
-	wget -p http://download.tensorflow.org/models/object_detection/ssd_mobilenet_v1_coco_11_06_2017.tar.gz --directory-prefix /tmp
-	tar -C ml/object-detection --strip-components 1 -xvf /tmp/download.tensorflow.org/models/object_detection/ssd_mobilenet_v1_coco_11_06_2017.tar.gz ssd_mobilenet_v1_coco_11_06_2017/frozen_inference_graph.pb
+opencv-base:
+	$(PACHCTL) create repo images
+	$(PACHCTL) create pipeline -f opencv/edges.json
+	$(PACHCTL) create pipeline -f opencv/montage.json
 
+opencv: opencv-base
+	$(PACHCTL) put file images@master -i opencv/images.txt
+	$(PACHCTL) put file images@master -i opencv/images2.txt
 
-object-detection-images-repo: ml/object-detection/frozen_inference_graph.pb
-	pachctl create repo images
+opencv-delete:
+	$(PACHCTL) delete pipeline montage
+	$(PACHCTL) delete pipeline edges
+	$(PACHCTL) delete repo images
 
-object-detection-no-images-repo: ml/object-detection/frozen_inference_graph.pb
-
-object-detection-demo: object-detection-no-images-repo
-	pachctl create repo training
-	pachctl put file training@master:frozen_inference_graph.pb -f ml/object-detection/frozen_inference_graph.pb
-	pachctl put file images@master:dogs.jpg -f ml/object-detection/images/dogs.jpg
+ml/object-detection/frozen_inference_graph.pb:
+	$(WGET) -p http://download.tensorflow.org/models/object_detection/ssd_mobilenet_v1_coco_11_06_2017.tar.gz --directory-prefix /tmp
+	$(TAR) -C ml/object-detection --strip-components 1 -xvf /tmp/download.tensorflow.org/models/object_detection/ssd_mobilenet_v1_coco_11_06_2017.tar.gz ssd_mobilenet_v1_coco_11_06_2017/frozen_inference_graph.pb
 
-object-detection:  object-detection-images-repo object-detection-demo
-	pachctl create pipeline -f ml/object-detection/model.json
-	pachctl create pipeline -f ml/object-detection/detect.json
+OBJECT_MODEL_PIPELINE = pachctl create pipeline -f ml/object-detection/model.json
+OBJECT_DETECT_PIPELINE = pachctl create pipeline -f ml/object-detection/detect.json
```

Comment:
These are put in variables so you can override them in other targets, right? If so, I think it'd be more straight-forward to not use variables here, and just copy/paste as necessary.

Reply:
They're overridden in Makefile.testing.  I'm doing it as an exercise in DRY.  I note that I'm not using $(PACHCTL) here, either!

Reply:
Yeah DRY as a general principle makes sense, but I think it may be overdoing it here

Reply:
I really want to keep this, as it's easier to build variations on the pipelines by overriding the variables in new targets. Take a look at the variation I do in Makefile.testing, for example. I'd need to dupllicate lines, possibly introducing errors in the ordering in which pipelines are created, if I don't have this structure. It enforces uniformity in the testing structure.

Reply:
This is what you have right now, between `Makefile` and `Makefile.testing`:

```
OBJECT_MODEL_PIPELINE = pachctl create pipeline -f ml/object-detection/model.json
OBJECT_DETECT_PIPELINE = pachctl create pipeline -f ml/object-detection/detect.json
object-detection-base: 
	@$(OBJECT_IMAGES)
	$(PACHCTL) create repo training
	@$(OBJECT_MODEL_PIPELINE)
	@$(OBJECT_DETECT_PIPELINE)

object-detection-testing-base: OBJECT_IMAGES = $(ECHO) "object-detection will be using opencv's images repo"
object-detection-testing-base: OBJECT_MODEL_PIPELINE = $(JQ) '$(STATS)|.pipeline.name="object-model"'  ml/object-detection/model.json | $(PACHCTL) create pipeline -f - >/dev/null
object-detection-testing-base: OBJECT_DETECT_PIPELINE = $(JQ) '$(STATS)|$(DETECTMODEL)'  ml/object-detection/detect.json | $(PACHCTL) create pipeline -f - >/dev/null
object-detection-testing-base:  object-detection-base
```

This is what it would look like without vars:

```
object-detection-base: 
    pachctl create repo images
    pachctl create repo training
    pachctl create pipeline -f ml/object-detection/model.json
    pachctl create pipeline -f ml/object-detection/detect.json

object-detection-testing-base: 
    echo "object-detection will be using opencv's images repo"
    pachctl create repo training
    jq '$(STATS)|.pipeline.name="object-model"'  ml/object-detection/model.json | pachctl create pipeline -f - >/dev/null
    jq '$(STATS)|$(DETECTMODEL)'  ml/object-detection/detect.json | pachctl create pipeline -f - >/dev/null
```

I think the latter is more straight-forward to understand, because there's less indirection, and it relies on less makefile magic. While DRY is generally a good principal to follow, it has been balanced with other principals.

Is there some other context I'm missing here? e.g. are you intending to create more `Makefile`s in the future that override these targets?

Reply:
I've kept this in my latest commit for the reasons above.

Reply:
Yes, in the future there will be more targets and more makefiles.  So, for example, if I wanted to create a version where stats is turned  *off* for just one of the pipelines, it's two lines of code.

```
object-detection-testing-nostats: OBJECT_DETECT_PIPELINE = blah blah blah
object-detection-testing-nostats: object-detection-testing-base
```